server-spi-private/src/main/java/org/keycloak/device/DeviceActivityManager.java

```diff
@@ -0,0 +1,215 @@
+            return;
+        } 
+        UserDeviceStore deviceProvider = getDeviceProvider(session);
+        
+        if (deviceProvider == null) {
+            return;
+        }
+        
+        UserModel user = userSession.getUser();
+        
+        if (user == null) {
+            return;
+        }
+        
+        if (session.getAttribute(DEVICE_UPDATED_FLAG) == null) {
+            DeviceModel current = getCurrentDevice(session);
+
+            if (current != null) {
+                DeviceModel device = getUserDevice(userSession, current, deviceProvider);
+
+                if (device == null) {
+                    registerDevice(userSession, current, deviceProvider);
+                } else {
+                    updateDevice(userSession, current, device, deviceProvider);
+                }
+
+                session.setAttribute(DEVICE_UPDATED_FLAG, true);
+            }
+        }
+    }
+    
+    private static UserDeviceStore getDeviceProvider(KeycloakSession session) {
+        UserProvider userProvider = session.users();
+
+        if (userProvider instanceof UserDeviceStore) {
+            return (UserDeviceStore) userProvider;
+        }
+
+        return null;
+    }
+
+    private static DeviceModel getUserDevice(UserSessionModel userSession, DeviceModel current, UserDeviceStore deviceProvider) {
+        List<DeviceModel> devices = deviceProvider.getDevices(userSession.getUser());
+
+        if (devices.isEmpty()) {
+            return null;
+        }
+
+        String attached = userSession.getNote(DeviceModel.DEVICE_ID);
+
+        if (attached != null) {
+            DeviceModel device = deviceProvider.getDeviceById(userSession.getUser(), attached);
+
+            if (device != null) {
+                // TODO: if current device is different than the one associated with the session, we just return what is in session. 
+                //  See how to handle this situation
+                return device; 
+            }
+        } else {
+            for (DeviceModel existing : devices) {
+                if (isSameDevice(userSession, current, existing)) {
+                    return existing;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static void registerDevice(UserSessionModel userSession, DeviceModel current, UserDeviceStore deviceProvider) {
+        current.setSession(userSession.getId());
+        deviceProvider.addDevice(userSession.getRealm(), userSession.getUser(), current);
+        if (userSession.getNote(DeviceModel.DEVICE_ID) == null) {
+            userSession.setNote(DeviceModel.DEVICE_ID, current.getId());
+        }
+    }
+
+    private static void updateDevice(UserSessionModel userSession, DeviceModel current, DeviceModel device, UserDeviceStore deviceProvider) {
+        if (!device.getBrowser().contains(current.getBrowser())) {
+            device.addBrowser(current.getBrowser());
+        }
+        device.setIp(current.getIp());
```

Comment:
For "newSession" and "lastAccess" it may be good to not have them in the DB at all, but maybe just in infinispan.

Reply:
I can remove the reference to the session, but I can't remove last access. The reason is that we need to able to show when a device was last accessed to authenticate regardless of the session being active or not.

Reply:
ok, thanks. For the offline sessions, we have a mechanism, which allows to update "lastAccess" time in a batches. So you have one DB request, which is triggered either periodically every 10 seconds or when 100 sessions are ready to be written (depends on what is faster). Same is used for update "lastAccess" time of sessions to cross-dc. Maybe this can be used as well to improve performance?

Reply:
Sounds interesting. Currently, you have a write whenever the device's `lastAccess` differs from the session it is attached to. For regular sessions, that means writing to the DB every time you set the last session refresh. I've looked at how you did for offline sessions and I think we can have a similar solution and avoid this problem.

In addition to `lastAccess` we also update the IP address and the browser. But these are only updated when they actually change, not every time. So I think we can keep them that way and focus on avoiding the `lastAccess` problem. WDYT ?